### ISSUE TITLE
Asp.Net Core IoC registration

### ISSUE THREAD

**Issue by nhwilly** [JeremySkinner/FluentValidation#465]
Really love this library.  Thank you for it.

I'm struggling getting dependency injection to work.  In the [ASP.NET Core integration](https://github.com/JeremySkinner/FluentValidation/wiki/i.-ASP.NET-Core-integration) page, you mention:

> To automatically register all validators with the container, you can use the AddFromAssemblyContaining method to point the registrar at the assembly containing your validator classes:

But I can't find `AddFromAssemblyContaining`.  Is that typo, or am I missing something?  

And when I use the `fv => fv.RegisterValidatorsFromAssembly(Assembly.GetEntryAssembly())` approach, I get errors because the container can't resolve the type, but here's a shot of my services:

![image](https://cloud.githubusercontent.com/assets/7560842/25868126/cf51819e-34c9-11e7-8382-9912e79bd2bf.png)

I see a `ValidatorFactory` that's registered as singleton, but my validator is a transient.

And a related question from [Creating a Validator](https://github.com/JeremySkinner/FluentValidation/wiki/b.-Creating-a-Validator):

> Because of this, it's recommended that you use validator instances as singletons- once instantiated they should be cached and reused, rather than being instantiated multiple times. 

I figure I must be doing something wrong, but I can't figure out what.

What does work is this:
`services.AddSingleton(typeof(InvitationValidator), new InvitationValidator())` but I don't want to do it one by one just because I'm missing something obvious.

Thanks in advance for any pointers.

**Comment by nhwilly** [comment]
This might be a moot point.  Everything coming into my site is a DTO.  Everything I want to validate with FV is an entity.  I was planning on using `DataAnnotations` on my DTO's, validate them with an action filter and let the pipeline toss them before they even got to my controller.

But I noticed that doing the `RegisterValidators...` call clears all the other model validators, which will zap my DTO validations.

Maybe I should try to get FV working in my action filter and dump all the other validation methods anyway.

I'd still like to figure out why this doesn't work.

**Comment by JeremySkinner** [comment]
So a few things:

Yes the method is called RegisterValidatorsFromAssembly (or RegisterValidatorsFromAssemblyContaining[T]). These methods register the validators with the container so that if you try and resolve the interface (i.e. IValidator[T]) then the correct implementer will be returned. Note that you must resolve using the interface not the concrete type (which is what the default validator factory does internally). What's the error you're seeing?

The validators are registered as transient. Singleton is recommended for performance but this can lead to issues with scoping of dependencies especially in a web context for users that are less experienced with how DI works. Transient is simpler, and FV now has caching that mitigates the worst of the performance issues. 

There is actually a bug in 6.4 where the validator factory is singleton when it should be request-scoped (this is fixed in the 7.0 betas). 

7.0 also fixes the issues you're seeing with other validator providers being wiped out. Note that although you can mix FV and other providers in the same project, you can't mix them for the same type. (FV is actually primarily designed for validating DTOs/input-models, not entities).

**Comment by nhwilly** [comment]
Thanks for the quick response.

I'll skip forward to the 7.0 beta as I have a while before I actually try to put anything in production.  That will take care of the factory problem and the clearing of other validators.  I didn't plan on mixing validations in the same type, so no problem there.

And I'm ok with them being transient if they're cached.  I was just trying to follow your advice.

And yes, this is a web project and yes, I'm out of my depth on DI, to be sure (although I love it already).

The error I'm seeing is that it can't resolve the type (I've closed the project for the moment).  And I did switch to trying to use an `IValidator<T>` and it did the same thing.

I am going to install the beta, recheck my details to make sure I haven't missed anything and let you know.

Thanks again.

**Comment by nhwilly** [comment]
Well, I'm losing my mind.  It seems to be working great now.  

I hate when that happens.  

I changed the expression to : `.AddFluentValidation(fv => fv.RegisterValidatorsFromAssemblyContaining<Startup>()); ` and then thought I'd run it before I installed the beta, just to be certain.  Boom.  Works.  Sorry.

I'll install Beta 7 just to address the other issues and if I have any, I'll triple check before I report them.

Thank you very much for your help.
